File: backend/src/main/java/com/giapha/model/entity/TimeCapsule.java
package com.giapha.model.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class TimeCapsule {
    private UUID id;
    private UUID familyId;
    private UUID creatorId;
    private String title;
    private String content;
    private String mediaUrl;
    private UUID recipientMemberId;
    private LocalDate unlockDate;
    private String unlockCondition;
    private String unlockEvent;
    private Boolean isOpened;
    private OffsetDateTime openedAt;
    private UUID openedBy;
    private OffsetDateTime createdAt;
}

File: frontend/src/types/time-capsule.ts
import type { FamilyMember } from './family';

/**
 * Status of a time capsule as returned by the backend.
 * - `sealed`: still locked, not yet openable
 * - `available`: unlock conditions met, ready to be opened
 * - `opened`: already opened by an authorized user
 */
export type TimeCapsuleStatus = 'sealed' | 'available' | 'opened';

export interface TimeCapsule {
  id: string;
  familyId: string;
  creatorId: string;
  title: string;
  content?: string;
  mediaUrl?: string;
  recipientMemberId?: string;
  unlockDate?: string;
  unlockCondition?: 'DATE' | 'EVENT' | 'MANUAL';
  unlockEvent?: string;
  isOpened?: boolean;
  openedAt?: string;
  openedBy?: string;
  createdAt?: string;
}

/** Wrapper for a single capsule returned alongside metadata. */
export interface TimeCapsuleEntry {
  capsule: TimeCapsule;
  creator?: { id?: string; fullName?: string };
  recipient?: { id?: string; fullName?: string } | null;
  daysUntilUnlock?: number;
  isUnlockable?: boolean;
}

export interface TimeCapsuleListResponse {
  capsules: TimeCapsuleEntry[];
}

export interface OpenTimeCapsuleResponse {
  capsule: TimeCapsule;
  content?: string;
}

export interface CreateTimeCapsuleRequest {
  familyId: string;
  title: string;
  content?: string;
  mediaUrl?: string;
  recipientMemberId?: string;
  /** ISO date (YYYY-MM-DD). Must be in the future. */
  unlockDate: string;
  unlockCondition: 'DATE' | 'EVENT' | 'MANUAL';
  unlockEvent?: string;
}

export interface CountdownInfo {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
  isUnlocked: boolean;
  totalMs: number;
}